src/utils/workspace-context-utils.ts

```diff
@@ -0,0 +1,117 @@
+  workspace: '',
+  setWorkspace: () => {},
+  workspaces: [],
+  workspacesLoaded: false,
+  lastUsedWorkspace: getActiveWorkspace(),
+});
+
+export const WorkspaceProvider = WorkspaceContext.Provider;
+
+export const useWorkspace = () => getActiveWorkspace();
+export const useLastUsedWorkspace = () => React.useContext(WorkspaceContext).lastUsedWorkspace;
+
+export const useActiveWorkspace = (): WorkspaceContextData => {
+  const lastUsedWorkspace = useLastUsedWorkspace();
+  const navigate = useNavigate();
+  const setActiveNamespace = useSetNamespace();
+  const [, workspaceFromUrl = ''] = window.location.pathname.match(workspacePathMatcher) || [];
+  const [workspace, setWorkspace] = React.useState<string>(getActiveWorkspace);
+  const [workspaces, setWorkspaces] = React.useState<any>([]);
+  const [workspacesLoaded, setWorkspacesLoaded] = React.useState<boolean>(false);
+
+  const getDefaultNsForWorkspace = React.useCallback((allWorkspaces, currentWorkspace) => {
+    const obj = allWorkspaces?.find((w) => w.metadata.name === currentWorkspace);
+    return obj?.status?.namespaces.find((n) => n.type === 'default');
+  }, []);
+
+  React.useEffect(() => {
+    if (workspace && workspaces?.length > 0) {
+      setActiveWorkspace(workspace);
+      const ns = getDefaultNsForWorkspace(workspaces, workspace);
+      if (ns) {
+        localStorage.setItem('activeNamespace', ns.name);
+        setActiveNamespace(ns.name);
+      }
+    }
+  }, [getDefaultNsForWorkspace, setActiveNamespace, workspace, workspaces]);
+
+  React.useEffect(() => {
+    let unmounted = false;
+    const fetchWorkspaces = async () => {
+      let allWorkspaces;
+      try {
+        setActiveWorkspace(''); // to fetch root level workspaces
+        allWorkspaces = await k8sListResourceItems({
+          model: WorkspaceModel,
+        });
+      } catch (e) {
+        // eslint-disable-next-line no-console
+        console.error('Error fetching workspaces', e);
+      }
+
+      if (unmounted) {
+        return;
+      }
+
+      let ws: string;
+      if (Array.isArray(allWorkspaces)) {
+        const workspaceNames = allWorkspaces.map((dataResource) => dataResource.metadata.name);
+        //Todo: using first workspace untill we get default workspace type in api response.
+
+        const isWorkspaceAvailable = (wsName: string) => workspaceNames.includes(wsName);
+
+        ws = isWorkspaceAvailable(workspaceFromUrl)
+          ? workspaceFromUrl
+          : isWorkspaceAvailable(lastUsedWorkspace)
+          ? lastUsedWorkspace
+          : workspaceNames[0];
+      } else if (allWorkspaces?.metadata?.namespace) {
+        ws = allWorkspaces.metadata.namespace;
+      }
+
+      if (ws) {
+        setWorkspace(ws);
+        setWorkspaces([...allWorkspaces]);
+        setWorkspacesLoaded(true);
+        const wsPath = generatePath('/stonesoup/workspaces/:ws/applications', { ws });
+        !window.location.pathname.includes(wsPath) && navigate(wsPath);
```

Comment:
This is causing overview page to immediately navigate to applications page on load.